design-docs/improved-console.md

```diff
@@ -0,0 +1,446 @@
+ - `ProgressLabel getStatusBar()` (inherited from `Console`)
+ - `MultiStatusTextArea getProgressArea()`
+ 
+An `StatusBarFormatter` has operations:
+ - `String format(ProgressOperation operation)`
+
+`MultiStatusTextArea` extends `TextArea` with the following operations:
+ - `void update(ProgressOperations operations)`
+ - `void setStatusBarFormatter(StatusBarFormatter formatter)` 
+
+> NOTE: Plugins can provide progress indication by publishing `BuildOperations` (separate effort)
+
+Implementation is similar to `ConsoleBackedProgressRenderer` with throttling and a data
+structure for storing recent updates.
+
+Warnings and errors are displayed in the main area with the same format and styling as today.
+
+`FixedSizeMultiStatusTextArea` would be an implementation of `MultiStatusTextArea` that
+initializes a `List<LinkedList<ProgressOperation>>` of fixed size (defaults to `org.gradle.workers.max`
+ under `--parallel` or 1 otherwise). It would keep track of `ProgressOperation` chains 
+in-progress to prevent having more in-progress operations than max workers displayed.
+
+> NOTE: We must avoid use of save/restore cursor position unless we switch to a 
+[terminfo](https://en.wikipedia.org/wiki/Terminfo)-based Console handling instead of using ANSI.
+
+User guide is updated with new output examples.
+    
+### Test Coverage
+* A terminal size with fewer than max parallel operations displays (rows - 2) operations. 
+ That is, the progressArea never grows taller than the console height - 2 rows.
+* Operation status lines are trimmed at (cols - 1)
+* `System.in` and SystemConsole I/O happens on the mainArea, which is unaffected by other areas
+* Renders well on default macOS Terminal, default Windows Console, Cygwin, PowerShell and
+ common Linux Terminals
+* Console shows idle workers under `--continuous` build
+
+## Story: Display overall build progress as a progress bar
+Render ProgressEvents with build completeness information as a progress bar.
+ 
+Intended to give the user a very fast way of telling whether the build will be finished soon or not.
+
+### User-visible Changes
+We can visually represent complete and un-started tasks of build using colorized output:
+
+`###############>                       > 40% Building`
+
+On ANSI terminals, we can use empty spaces with different background colors for the "bars".
+The progress bar shall be red given a failing build.
+
+Note that the suffix at the end of the progress bar should be the progress for the current
+Gradle execution phase.
+
+**ASCII-based options (some options use extended ASCII):**
+```
+[##                        ] 6% Building
+[‡‡‡‡                      ] 12% Building
+###########>               > 33% Building
+‹===================       › 60% Building
+«==========================» 100% BUILD SUCCESS
+```
+
+### Implementation
+`DefaultGradleLauncherFactory` registers a listener `BuildProgressBarListener` (similar to `BuildProgressFilter`)
```

Comment:
Yes, they're all build operations already. That's the point: don't add another ad hoc mechanism, instead add a general mechanism and _reuse_ it for configuration, test execution, native compilation, dependency resolution, etc. It's the same amount of effort to add the general mechanism as it is to add the ad hoc mechanism.